### ISSUE TITLE
Fix MV dtests with tablets enabled

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17038]
This issue is about dtests in materialized_views_test.py  which currently fail with the `tablets` experimental feature enabled, _beyond_ the tests that already fail with consistent topology changes enabled (see https://github.com/scylladb/scylladb/issues/17021 for those). 

I'll try to update this text as my understanding unfolds, as we have additional issues and PRs to refer to, and so on.

Failures due to self-pairing no longer happening with tablets (see also https://github.com/scylladb/scylladb/issues/17043). **Fixed** in https://github.com/scylladb/scylla-dtest/pull/3909:
* `test_repair_mv`
* `test_simple_repair`
* `test_complex_repair`
* `test_really_complex_repair`

The following test passes, but complains about unexpected error from the Raft systems while one of the nodes is temporarily down, or when it goes down during tablet migration - see details in https://github.com/scylladb/scylladb/issues/17290. **Fixed** in https://github.com/scylladb/scylla-dtest/pull/4082.
* `test_mv_populating_from_existing_data_during_node_restart` 

The following test because it brings down one node permanently, went into an endless loop of error messages during "tablet migration", and the test never completes. This was https://github.com/scylladb/scylladb/issues/16527#issuecomment-1938883123, and already fixed. Then the test sometimes fails  on an unexpected error in the log in deleting a non-empty hints directory - issue #17023. **Fixed** in #18377
* `test_mv_populating_from_existing_data_during_node_remove`

The following tests had problems both a decomissioned node fails to shut down cleanly at the end of the decommission because of a failing service-levels shutdown process: I opened an issue https://github.com/scylladb/scylladb/issues/17293 - which was **fixed** in https://github.com/scylladb/scylladb/pull/17548:
* `test_mv_resurrected_rows_after_decommission_interrupt[without_rbno]`
* `test_mv_resurrected_rows_after_decommission_interrupt[with_rbno]`
* `test_mv_populating_from_existing_data_during_node_decommission` 

(The first two tests also have consistent-topology-changes problems, see :point_right: https://github.com/scylladb/scylladb/issues/17026 and fixed https://github.com/scylladb/scylladb/issues/17027)
(The second also now failed on #17962, a text bug **fixed** by https://github.com/scylladb/scylla-dtest/pull/4193 )
The following test seemed to hang on a massive view building (building 50 views in parallel), which I don't remember happening without tablets. Opened issue https://github.com/scylladb/scylladb/issues/17297. It was **fixed** by some other fix to tablets (maybe https://github.com/scylladb/scylladb/pull/17846).
* `test_mv_alter_with_synchronous_updates`

The following test complains that "View building didn't start in 20 seconds" - I opened #18427. Fixed in https://github.com/scylladb/scylla-dtest/pull/4580  :point_right:  NOT YET MERGED. In the past it also complained about unexpected errors in the log but it seems it no longer does.
* `test_do_not_finish_view_building_with_hints` 

Additional failing tests:
* `test_interrupt_build_process_test[colocated_view_replicas]` - "View ks.t_by_v not built" and also a **HUGE** number of raft errors (similar to above case with huge number)
* `test_base_replica_repair` - ?

I need to investigate these failures. Some of them that have to do with unexpected ERROR-level logging will need to be solved by the raft topology team (CC @kbr-scylla). They can also be ignored in dtest, but we should ask ourselves if these are really ERROR-level logging events, or not, before just taking the easy path out and changing the test to ignore them.

**Comment by nyh** [comment]
As I suspected, `test_repair_mv` (and probably the other repair tests listed above) assumes that because they used RF=3 N=3, one can read from the view right after finishing to write to the base and be able to read everything. With tablets, this assumption is no longer correct, so the test sometimes, depending on timing, fails (it also sometimes succeeds, unsurprisingly).

The fix can to either enable synchronous_updates on the view in this test, or to add the missing "eventually".
It's easier for me to just add synchronous_updates - and it will also be less risky in the sense that I might miss one of the reads that needs an "eventually".

The downside of synchronous_updates is that will not work on Cassandra, but I don't think we ever cared and don't need to start caring now (these tests would have failed on Cassandra too because of the missing eventually).

**Comment by mykaul** [comment]
Isn't synchronous view a different feature that needs to be tested, and 'eventually' is the 'default' mode we have and need to support?

**Comment by nyh** [comment]
> Isn't synchronous view a different feature that needs to be tested, and 'eventually' is the 'default' mode we have and need to support?

Yes, sort of. In theory, this test was supposed to be about **repair**, so could have used other features freely, and moreover the part of the test which fails is the setup - it's not even the real part of the test. But you're right that it's better to not depend on synchronous view updates if we don't have to. I won't be using it anyway, I switched to using eventually.

**Comment by mykaul** [comment]
@nyh - what's the latest here?

**Comment by kbr-scylla** [comment]
> I need to investigate these failures. Some of them that have to do with unexpected ERROR-level logging will need to be solved by the raft topology team (CC @kbr-scylla).

If there are any remaining problems like that --- please open issues and assign them to me.

**Comment by bhalevy** [comment]
@nyh, is https://jenkins.scylladb.com/view/master/job/scylla-master/job/tablets/job/dtest-release-with-tablets/10/testReport/materialized_views_test/TestInterruptBuildProcess/Run_Dtest_Parallel_Cloud_Machines___HeavyDtest___heavy_split000___test_interrupt_build_process_test_colocated_view_replicas_/ even valid with tablets?
Does colocated view replicas means that they are expected to be on the same node?
We don't guarantee that with tablets AFAIK.
@tgrabiec ?

**Comment by tgrabiec** [comment]
> Does colocated view replicas means that they are expected to be on the same node?
> We don't guarantee that with tablets AFAIK.
> @tgrabiec ?

Tablet load balancer definitely doesn't maintain colocation between base and view tablets in terms of location on the same node/shard.

**Comment by avikivity** [comment]
> > Does colocated view replicas means that they are expected to be on the same node?
> > We don't guarantee that with tablets AFAIK.
> > @tgrabiec ?
> 
> Tablet load balancer definitely doesn't maintain colocation between base and view tablets in terms of location on the same node/shard.

It should for local views (where the partition key is shared). The view tablets should be subordinate to the base tablets (split decision, migration, everything should match exactly).

**Comment by swasik** [comment]
@yarongilor - you are the last one who modified the state of this issue by reopening it. Could you comment what is the status of it and if we have some plan for working on it?

**Comment by yarongilor** [comment]
> [@yarongilor](https://github.com/yarongilor) - you are the last one who modified the state of this issue by reopening it. Could you comment what is the status of it and if we have some plan for working on it?

sorry, i think it was closed by a mistake.
The issue of disabling self-pairing (colocation between base and view tablets) is not resolved. i only adjusted dtests to workaround it, like https://github.com/scylladb/scylla-dtest/pull/4560. According to above comments like Avi's, there may be fixes to do here or in scylla. So it's for @nyh to update.

**Comment by mlitvk** [comment]
MV tests that are currently skipped with tablets:
* `test_remove_node_during_mv_insert_3_nodes`
The test scenario seems illegal with tablets. it starts with 3 nodes, rf=3, and removes a node. removing the node fails ("Unable to find new replica for tablet"). There is another test case with 4 nodes that covers this.
* `test_add_dc_after_mv_simple_replication`
The test passes but it says that it's redundant because SimpleStrategy is not used, and there is another case that covers the network replication case.
* `test_lwt`
LWT is currently not supported with tablets:
https://github.com/scylladb/scylladb/issues/18068
* `test_do_not_finish_view_building_with_hints[colocated_view_replicas]`
* `test_interrupt_build_process_test[colocated_view_replicas]`
* `test_complex_repair`
* `test_really_complex_repair`
These 4 cases assume colocation of base and view replicas which we don't have currently with tablets.
https://github.com/scylladb/scylladb/issues/17043
* `test_interrupt_build_process_with_resharding_low_to_half_test`
* `test_interrupt_build_process_with_resharding_half_to_max_test`
* `test_interrupt_build_process_with_resharding_max_to_half_test`
* `test_interrupt_build_process_with_resharding_half_to_low_test`
* `test_interrupt_build_process_and_resharding_low_to_half_test`
* `test_interrupt_build_process_and_resharding_half_to_max_test`
* `test_interrupt_build_process_and_resharding_max_to_half_test`
* `test_interrupt_build_process_and_resharding_half_to_low_test`
Resharding is currently not supported with tablets:
https://github.com/scylladb/scylladb/issues/16739

This test was previously skipped and now enabled:
* `test_add_dc_after_mv_network_replication`
Enabled in https://github.com/scylladb/scylla-dtest/pull/5493

The other tests pass with tablets

**Comment by piodul** [comment]
> * `test_remove_node_during_mv_insert_3_nodes`
>       The test scenario seems illegal with tablets. it starts with 3 nodes, rf=3, and removes a node. removing the node fails ("Unable to find new replica for tablet"). There is another test case with 4 nodes that covers this.

Yes, it looks like it is just not allowed. The error message is not perfect, but there is an issue about improving it: https://github.com/scylladb/scylladb/issues/20417.

The test should be skipped as it is not, but there should be a comment which explains it.

```python3
    # cassandra-stress enables tables due to using NetworkTopologyStrategy
    # Covered by test_remove_node_during_mv_insert_4_nodes when tablets are enabled
```

I don't find the current comment (shown above) to be too clear, let's improve it.

> * `test_add_dc_after_mv_simple_replication`
>   The test passes but it says that it's redundant because SimpleStrategy is not used, and there is another case that covers the network replication case.

Looks good, the test should be kept marked as skipped for tablets, the current explanation in the comment above the test looks good.

----

@mlitvk About the rest - let's make sure that they are marked appropriately with `issue_open` so that they will get re-enabled when the associated issues get closed. Some of them already are but some others seem to be missing the right mark.

After this is done, let's close this issue - no need to keep it additionally open, tests will get automatically re-enabled after we fix the relevant issues for each test.
